python/ml_wrappers/model/predictions_wrapper.py

```diff
@@ -0,0 +1,199 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# ---------------------------------------------------------
+
+"""Defines classes to wrap the training/test data and the corresponding predictions from the model."""
+
+from typing import Optional
+
+import numpy as np
+import pandas as pd
+
+TARGET = 'target'
+
+
+class ModelWrapperPredictions:
+    """Model wrapper to wrap the samples used to train the models
+    and the predictions of the model. This wrapper is useful when
+    it is not possible to load the model.
+    """
+    def __init__(self, test_data: pd.DataFrame, y_pred: np.ndarray):
+        """Creates an ModelWrapper object.
+
+        :param test_data: The data that was used to train the model.
+        :type test_data: pd.DataFrame
+        :param y_pred: Predictions of the model.
+        :type y_pred: np.ndarray
+        """
+        if not isinstance(test_data, pd.DataFrame):
+            raise DataValidationException(
+                "Expecting a pandas dataframe for test_data")
+        if not isinstance(y_pred, np.ndarray):
+            raise DataValidationException(
+                "Expecting a numpy array for y_pred")
+        if len(test_data) != len(y_pred):
+            raise DataValidationException(
+                "The number of instances in test data "
+                "do not match with number of predictions")
+        self._combined_data = test_data.copy()
+        self._combined_data[TARGET] = y_pred
+
+    def _get_filtered_data(
+            self, query_test_data_row: pd.Series) -> pd.DataFrame:
+        """Return the filtered data based on the query data.
+
+        :param query_test_data_row: The query data instance.
+        :type query_test_data_row: pd.Series
+        :return: The filtered dataframe based on the values in query data.
+        :rtype: pd.DataFrame
+        """
+        queries = []
+        for column_name, column_data in query_test_data_row.iteritems():
+            if isinstance(column_data, str):
+                queries.append("`{}` == '{}'".format(
+                    column_name, column_data))
+            else:
+                queries.append("`{}` == {}".format(
+                    column_name, column_data))
+
+        queries_str = '(' + ') & ('.join(queries) + ')'
+        filtered_df = self._combined_data.query(queries_str)
+
+        if len(filtered_df) == 0:
+            raise EmptyDataException(
+                "The query data was not found in the combined dataset")
+
+        return filtered_df
+
+    def predict(self, query_test_data: pd.DataFrame) -> np.ndarray:
+        """Return the predictions based on the query data.
+
+        :param query_test_data: The data for which the predictions need to
+            be returned.
+        :type query_test_data: pd.DataFrame
+        :return: Predictions of the model.
+        :rtype: np.ndarray
+        """
+        if not isinstance(query_test_data, pd.DataFrame):
```

Comment:
Looks great.  Did you want to use KNN instead, with nn=1, or keep the filtering logic here?

Reply:
The Knn could be used to replace the current search based approach. But there are few things to consider:-

1. We need to transform the string features as knn cannot take string features https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html
2. We need to think how we want to model the problem as a classification problem with as many classes as number of instances of test data. We need to do more testing if knn could work with such high number of classes. 

Maybe we can merge this and do a separate investigation on using knn inside of the search.